lib/src/main/java/org/hyperagents/jacamo/artifacts/yggdrasil/YggdrasilThingArtifact.java

```diff
@@ -0,0 +1,83 @@
+package org.hyperagents.jacamo.artifacts.yggdrasil;
+
+import java.util.Optional;
+import java.io.IOException;
+
+import cartago.LINK;
+
+import org.hyperagents.jacamo.artifacts.wot.ThingArtifact;
+
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.client5.http.fluent.Request;
+
+/**
+ * Extension to the ThingArtifact class that adds Yggdrasil-specific WebSub
+ * support.
+ * YggdrasilThingArtifact is a subclass of ThingArtifact and provides additional
+ * functionality
+ * for registering WebSub to a Yggdrasil node.
+ *
+ * Contributors:
+ * - Andrei Ciortea (author), Interactions-HSG, University of St. Gallen
+ *
+ */
+public class YggdrasilThingArtifact extends ThingArtifact {
+
+    @Override
+    public void init(String url) {
+        super.init(url);
+        exposeWebSubIRIs(url);
+    }
+
+    @LINK
+    public void onNotification(Notification notification) {
+        log("The state of this ThingArtifact has changed: " + notification.getMessage());
+
+        String obsProp = notification.getMessage();
+        String functor = obsProp.substring(0, obsProp.indexOf("("));
+        String[] params = obsProp.substring(obsProp.indexOf("(") + 1, obsProp.length() - 1)
+                .split(",");
+
+        if (this.hasObsPropertyByTemplate(functor, (Object[]) params)) {
+            this.updateObsProperty(functor, (Object[]) params);
+        } else {
+            this.defineObsProperty(functor, (Object[]) params);
+        }
+    }
+
+    /*
+     * Registers for WebSub to an Yggdrasil node. This is not a generic
```

Comment:
Ideally, we should improve the implementation such that it is no longer specific to Yggdrasil — but works for any ThingArtifact using WebSub.